exercises/concept/locomotive-engineer/.docs/introduction.md

```diff
@@ -0,0 +1,419 @@
+=> [1, 2, 3, "Hello", "Mars"]
+```
+
+Usage of `**keyword_arguments`:
+
+```irb
+# This method is defined to take any number of keyword arguments
+
+>> def my_method(**keyword_arguments)= keyword_arguments
+
+# Arguments given to the method are packed into a dictionary
+
+>> my_method(a: 1, b: 2, c: 3)
+=> {:a => 1, :b => 2, :c => 3}
+```
+
+If the method defined does not have any defined parameters for keyword arguments(`**kwargs` or `<key_word>: <value>`) then the keyword arguments will be packed into a hash and assigned to the last parameter.
+
+```irb
+>> def my_method(a)= a
+
+>> my_method(a: 1, b: 2, c: 3)
+=> {:a => 1, :b => 2, :c => 3}
+```
+
+`*arguments` and `**keyword_arguments` can also be used in combination with one another:
+
+```ruby
+def my_method(*arguments, **keywword_arguments)
+  p args.sum
+  for (key, value) in kwargs.to_a
+    p key.to_s + " = " + value.to_s
+  end
+end
+
+
+my_method(1, 2, 3, a: 1, b: 2, c: 3)
+6
+"a = 1"
+"b = 2"
+"c = 3"
+```
+
+You can also write arguments before and after `*arguments` to allow for specific positional arguments.
+This works the same way as decomposing an array.
+
+```exercism/caution
+Arguments have to be structured in a specific order:
+
+`def my_method(<positional_args>, *args, <positional_args>, <key-word_args>, **kwargs)`
+
+If you don't follow this order then you will get an error.
+```
+
+```ruby
+def my_method(a, b, *arguments)
+  p a
+  p b
+  p arguments
+end
+
+my_method(1, 2, 3, 4, 5)
+1
+2
+[3, 4, 5]
+```
+
+You can write positional arguments before and after `*args`:
+
+```irb
+>> def my_method(a, *middle, b)= middle
+
+>> my_method(1, 2, 3, 4, 5)
+=> [2, 3, 4]
+```
+
+You can also combine positional arguments, \*arguments, key-word arguments and \*\*keyword_arguments:
+
+```irb
+>> def my_method(first, *many, last, a:, **keyword_arguments)
+     p first
+     p many
+     p last
+     p a
+     p keyword_arguments
+     end
+
+>> my_method(1, 2, 3, 4, 5, a: 6, b: 7, c: 8)
+1
+[2, 3, 4]
+5
+6
+{:b => 7, :c => 8}
+```
+
+Writing arguments in an incorrect order will result in an error:
+
+```ruby
+def my_method(a:, **kwargs, first, *args, last)
+  args
+end
+
+my_method(1, 2, 3, 4, a: 5)
+
+syntax error, unexpected local variable or method, expecting & or '&'
+... my_method(a:, **kwargs, first, *args, last)
+```
+
+### Decomposing into method calls
+
+You can use splat operator (`*`) to unpack an **array** of arguments into a method call:
+
+```ruby
+def my_method(a, b, c)
+  p c
+  p b
+  p a
+end
+
+numbers = [1, 2, 3]
+my_method(*numbers)
+3
+2
+1
+```
+
+You can also use double splat operator(`**`) to unpack a **hash** of arguments into a method call:
+
+```ruby
+def my_method(a:, b:, c:)
+  p c
+  p b
+  p a
+end
+
+numbers = {a: 1, b: 2, c: 3}
+my_method(**numbers)
+3
+2
+1
+```
+
+[args]: https://docs.ruby-lang.org/en/3.1/syntax/methods_rdoc.html#label-Array-2FHash+Argument
+[kwargs]: https://docs.ruby-lang.org/en/3.1/syntax/methods_rdoc.html#label-Keyword+Arguments
```

Comment:
We use full words for other reference links, not sure why we are using abbreviations for these two.